R/aggregate_by_vector.R

```diff
@@ -0,0 +1,173 @@
+
+    sampleid <- seq_along(unique(sample))
+    ## This will be needed for NUSE-type of quality control, but will
+    ## need to check for missing data as below.
+    ## se <- unique(summary(fit)$coefficients[sampleid, 'Std. Error'])
+
+    ## Take the sample coefficients ( = summarised expression values)
+    coef  <-  fit$coefficients[sampleid]
+    ## Sort the sample coefficients in the same way as the samplenames
+    ## of expression matrix. Puts NA for the samples without any
+    ## expression value
+    res <- coef[paste0('sample', colnames(x))]
+    names(res) <- colnames(x)
+    res
+}
+
+##' @title Return the Median Polish (Robust Twoway Decomposition) of a matrix
+##'
+##' @description
+##' Fits an additive model (two way decomposition) using Tukey's median
+##' polish procedure using [stats::medpolish()].
+##'
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param verbose Default is `FALSE`.
+##'
+##' @param ... Additional arguments passed to [stats::medpolish()].
+##'
+##' @return A `numeric` vector of length identical to `ncol(x)`.
+##'
+##' @author Laurent Gatto
+##'
+##' @family Quantitative feature aggregation
+##'
+##' @export
+##'
+##' @importFrom stats medpolish
+##'
+##' @examples
+##' x <- matrix(rnorm(30), nrow = 3)
+##' medianPolish(x)
+medianPolish <- function(x, verbose = FALSE, ...) {
+    medpol <- stats::medpolish(x, trace.iter = verbose, ...)
+    medpol$overall + medpol$col
+}
+
+##' @title Counts the number of features
+##'
+##' @description
+##' Returns the number of non-NA features in a features by sample
+##' matrix.
+##'
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param ... Currently ignored.
+##'
+##' @return A `numeric` vector of length identical to `ncol(x)`.
+##'
+##' @family Quantitative feature aggregation
+##'
+##' @export
+##'
+##' @author Laurent Gatto
+##'
+##' @examples
+##' m <- matrix(c(1, NA, 2, 3, NA, NA, 4, 5, 6),
+##'             nrow = 3)
+##' colCounts(m)
+##' m <- matrix(rnorm(30), nrow = 3)
+##' colCounts(m)
+colCounts <- function(x, ...)
+    colSums(!is.na(x))
+
+
+
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param INDEX A `vector` or `factor` of length `nrow(x)`.
+##'
+##' @param FUN A `function` to be applied to the subsets of `x`.
+##'
+##' @param ... Additional arguments passed to `FUN`.
+##'
+##' @return [aggregate_by_vector()] returns a new `matrix` of
+##'     dimensions `length(INDEX)` and `ncol(x), with `dimnames` equal
+##'     to `colnames(x)` and `INDEX`.
+##'
+##' @rdname aggregate
+##'
+##' @export
+aggregate_by_vector <- function(x, INDEX, FUN, ...) {
+    if (!inherits(x, "matrix"))
+        stop("'x' must be a matrix.")
+    if (!identical(length(INDEX), nrow(x)))
+        stop("The length of 'INDEX' has to be identical to 'nrow(x).")
+    INDEX <- factor(INDEX)
+    res <- tapply(
+        seq_len(nrow(x)),
+        INDEX,
+        FUN = function(i) FUN(x[i, , drop = FALSE], ...),
+        simplify = FALSE
+    )
```

Comment:
I would simplify this call by `lapply(split(...` (which is more or less what `tapply(..., simplify = FALSE)` is doing:
```suggestion
    FUN <- match.fun(FUN)
    res <- lapply(
        split(seq_len(nrow(x)), INDEX),
        FUN = function(i) FUN(x[i, , drop = FALSE], ...)
    )
```
It is less code and slightly improves the runtime:
```r
library("microbenchmark")
set.seed(1)

m <- matrix(sample(1000, 1000, repl = TRUE), nrow = 100)
i <- sample(100, 100, repl = TRUE)

aggregate_by_vector <- function(x, INDEX, FUN, ...) {
    FUN <- match.fun(FUN)
    res <- tapply(
        seq_len(nrow(x)),
        INDEX,
        FUN = function(i) FUN(x[i, , drop = FALSE]),
        simplify = FALSE
    )
    do.call(rbind, res)
}

aggregate_by_vector2 <- function(x, INDEX, FUN, ...) {
    fun <- match.fun(FUN)
    res <- lapply(
        split(seq_len(nrow(x)), INDEX),
        FUN = function(i) fun(x[i, , drop = FALSE])
    )
    do.call(rbind, res)
}

microbenchmark(
    tapply = aggregate_by_vector(m, i, colMeans),
    lapply = aggregate_by_vector2(m, i, colMeans),
    check = "identical"
)
# Unit: microseconds
#    expr     min      lq     mean  median       uq      max neval
#  tapply 516.905 553.142 621.4333 578.680 613.2660 2384.372   100
#  lapply 480.152 512.610 539.0358 531.765 560.4715  751.225   100
``

Reply:
Can't argue with a benchmark!